File: Python/textbook_track/chapter06/ba6e.py
# Find All Shared k-mers of a Pair of Strings
import sys

from collections import defaultdict

NUCLEOTIDE_COMPLEMENTS = {'A': 'T', 'C': 'G', 'T': 'A', 'G': 'C'}


def calc_reverse_complement(genome):
    return ''.join(map(NUCLEOTIDE_COMPLEMENTS.get, reversed(genome)))


def collect_k_mer_indices(genome, k):
    indices = defaultdict(list)
    for ix in range(len(genome)-k+1):
        k_mer1 = genome[ix:ix+k]
        indices[k_mer1].append(ix)
    return dict(indices)


def calc_shared_k_mers(s1, s2, k):
    k_mer_indices = collect_k_mer_indices(s1, k)
    shared_k_mer_indices = []
    for jy in range(len(s2)-k+1):
        k_mer = s2[jy:jy+k]
        reversed_k_mer = calc_reverse_complement(k_mer)
        indices_in_s1 = k_mer_indices.get(k_mer, []) + k_mer_indices.get(reversed_k_mer, [])
        for ix in indices_in_s1:
            shared_k_mer_indices.append((ix, jy))
    return shared_k_mer_indices


def main():
    reader = sys.stdin
    k = int(next(reader))
    s1 = next(reader).rstrip()
    s2 = next(reader).rstrip()
    result = calc_shared_k_mers(s1, s2, k)
    for x, y in result:
        print(f'({x}, {y})')


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter06/BA6E.scala
package TextbookTrack.Chapter06

object BA6E {
  private val NucleotideComplements: Map[Char, Char] = Map('A' -> 'T', 'C' -> 'G', 'T' -> 'A', 'G' -> 'C')

  private def calcReverseComplement(dna: String): String = dna.reverseIterator.map(NucleotideComplements).mkString

  private def collectKMerIndices(genome: String, k: Int): Map[String, List[Int]] =
    genome
      .sliding(k)
      .zipWithIndex
      .toList
      .groupMap{ case (kMer, _) => kMer }{ case (_, ix) => ix }

  def calcSharedKMers(s1: String, s2: String, k: Int): List[(Int, Int)] = {
    val kMerIndices: Map[String, List[Int]] = collectKMerIndices(s1, k)
    s2
      .sliding(k)
      .zipWithIndex
      .flatMap{ case (kMer, jy) =>
        val reversedKMer: String = calcReverseComplement(kMer)
        (kMerIndices.getOrElse(kMer, Nil) ::: kMerIndices.getOrElse(reversedKMer, Nil)).map(ix => (ix, jy))
      }
      .toList
  }


  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val k: Int = reader.next().toInt
    val s1: String = reader.next()
    val s2: String = reader.next()
    val result: List[(Int, Int)] = calcSharedKMers(s1, s2, k)
    result.foreach{ case (ix, jy) => println(s"($ix, $jy)") }
  }
}
